### PR TITLE
Artem/42/handle websocket connection loss

### PR DESCRIPTION
**Pull request by vertex451**
Goes in pair with https://github.com/orkusinc/api/pull/2829
## Fix Description
Call chain is next 
```
api/common/utils/GremlinClient() > orkusinc/gremlin/connection/Exec(req)
```
In Exec method we Send message and read it response.
When connection is lost, we receive EOF error - and then  call `pool.createSocket()` method
`pool.createSocket()` method has backoff logic of creating new socket connection.
After connection recover we goto Send message again to not to lose this request.

Also we call `pool.createSocket()` method while creating newService - so service will be created only after gremlin server is up

## How I have tested it 
```
kubectl delete deployment gremlinserver-deployment
make launch-kubernetes
kubectl apply -f kubernetes/configs/minikube/deployment-gremlinserver.yaml
waitForPods.go
```

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -10,39 +10,47 @@ import (
 	"strings"
 	"time"
 
+	orkuserr "github.com/orkusinc/api/common/errors"
```
Comment (achandak123) [inline connection.go]: Dont include orkus packages into vendor, as its supposed to be open source.

File: connection.go [hunk 2]
```diff
@@ -88,20 +102,31 @@ func (p *Pool) ExecQuery(query string) ([]byte, error) {
...
+start:
 	if err := websocket.Message.Send(conn.ws, requestMessage); err != nil {
+		orkuserr.PrintError(p.logger, "error while sending message", err)
 		return nil, err
 	}
-
 	data, err := conn.ReadResponse()
+	if err != nil {
+		orkuserr.PrintError(p.logger, "error while reading response", err)
+		if err.Error() == "EOF" { // EOF err we are getting on connection loss
```
Comment (achandak123) [inline connection.go]: We should use io.EOF instead.

File: connection.go [hunk 3]
```diff
@@ -10,39 +10,47 @@ import (
...
+func NewClient(logger log.Logger, urlStr string, origin string, maxConn []int, options ...OptAuth) (*Pool, error) {
 	pool := &Pool{
 		urlStr: urlStr,
 		origin: origin,
 		Auth:   options,
+		logger: logger,
+	}
+
+	// Check if connection is possible and close it
+	ws, err := pool.createSocket(0)
```
Comment (achandak123) [inline connection.go]: Why are we passing an argument to createSocket

File: connection.go (hunk 1 above)
Comment (vertex451) [reply connection.go]: Done

File: connection.go (hunk 3 above)
Comment (vertex451) [reply connection.go]: removed

File: connection.go [hunk 4]
```diff
@@ -4,45 +4,51 @@ import (
...
-	}
+func NewClient(logger log.Logger, urlStr string, origin string, maxConn []int, options ...OptAuth) (*Pool, error) {
 	pool := &Pool{
 		urlStr: urlStr,
 		origin: origin,
 		Auth:   options,
+		logger: logger,
 	}
+
+	// Check if connection is possible and close it
```
Comment (achandak123) [inline connection.go]: Why do we need to do this?

File: connection.go [hunk 5]
```diff
@@ -88,20 +99,31 @@ func (p *Pool) ExecQuery(query string) ([]byte, error) {
...
-	// Open a TCP connection
+start:
 	if err := websocket.Message.Send(conn.ws, requestMessage); err != nil {
+		p.logger.Log("error while sending message", err)
 		return nil, err
 	}
-
 	data, err := conn.ReadResponse()
+	if err != nil {
+		p.logger.Log("error while reading response", err)
```
Comment (achandak123) [inline connection.go]: In logger, the way it works is like key, value, so first argument is the key, second is the value, third is key, fourth is value and so on. Key cannot have spaces in it.

Comment (achandak123) [reply connection.go]: We should log this error only in the else case, wherein the error is not io.EOF(), as that one we are already handling.

**Review by achandak123 — CHANGES_REQUESTED** [review]
Some more comments

File: connection.go (hunk 4 above)
Comment (vertex451) [reply connection.go]: It is for initial check if gremlin is reachable on gremlin client creation
but the same is done in a loop few rows down - so I removed it

**Review by vertex451 — COMMENTED** [review]
Taken care

**Review by achandak123 — APPROVED** [review]

**Pull request merged**